---
title: "RNA-seq of GM-CSF+ and GM-CSF- CD4+ T cells"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/liubov/Documents/rnaseq_GSE122946/")
```

```{r, message=FALSE}
library(dplyr)
library(edgeR)
library(org.Hs.eg.db)
```

## Creating a merged table of expression counts

```{r}
dfs <- lapply(list.files("./htseq_counts/", full.names = TRUE),
              read.delim, header = FALSE)
samples = sapply(list.files("./htseq_counts/"),
                 function(x) unlist(strsplit(x, "[.]"))[1])
dfs <- mapply(function(x, y){colnames(x) <- c("gene", y); return(x)},
              dfs, samples, SIMPLIFY = FALSE)

merged_df <- Reduce(function(df1, df2) inner_join(df1, df2, by = "gene"), dfs)
# delete special counters, generated by htseq-count
merged_df <- subset(merged_df, !grepl("^__", gene))
dgelist <- DGEList(counts = merged_df[, 2:7], genes = merged_df[, 1])
nrow(dgelist$genes)
```

```{r, include=FALSE}
# write.table(merged_df, file = "./diff_expr/merged_counts.tsv", 
#             sep = "\t", row.names = FALSE)
```

```{r, include=FALSE}
# merged_df <- read.table("merged_counts.tsv", header = TRUE)
# merged_df <- subset(merged_df, !grepl("^__", gene))
# dgelist <- DGEList(counts = merged_df[, 2:7], genes = merged_df[, 1])
```

## Design

```{r}
donor <- factor(paste0("Donor", c("A", "A", "B", "B", "C","C")))
status <- factor(rep(c("CSF2_down", "CSF2_up"), 3))
design <- model.matrix(~ donor + status)
rownames(design) <- colnames(dgelist)
```

## Filtering and normalization

```{r}
keep <- filterByExpr(dgelist, group = status)
dgelist_filtered <- dgelist[keep, keep.lib.sizes = FALSE]
nrow(dgelist_filtered$genes)
```

```{r}
dgelist_filtered <- calcNormFactors(dgelist_filtered)
plotMDS(dgelist_filtered)
```

```{r}
dgelist_filtered <- estimateDisp(dgelist_filtered, design, robust = TRUE)
dgelist_filtered$common.dispersion
plotBCV(dgelist_filtered)
```

## Differential expression

```{r}
fit <- glmFit(dgelist_filtered, design)
lrt <- glmLRT(fit)
topTags(lrt)
```

```{r}
summary(decideTests(lrt, lfc = 1, p.value = 0.001))
```

```{r}
plotMD(lrt)
```

## Annotation

```{r}
go <- goana(lrt)
topGO(go, ont="MF", sort="Down", n=30, truncate=30)
```


```{r, include=FALSE}
nrow(dgelist$genes)
table(is.na(select(org.Hs.eg.db, dgelist$genes[,1], "ENTREZID", "UNIPROT"))[,2])
```

<!--- min 1 read in at least 3 samples

EdgeR (v.3.2.2): 
 * trimmed mean of M normalization
 * generalized linear model
 * likelihood ratio tests
 * transcript per million log2FC >= 1, FDR < 0.001 --->
